### ISSUE TITLE
DDP fails on SLURM clusters

### ISSUE THREAD

**Issue by Siddhant-Ray** [Lightning-AI/lightning#14022]
## 🐛 Bug

Using the Distributed DataParallel accelerator on SLURM managed GPU clusters fails.

### To Reproduce
```
File name test.py

import os

import torch
from torch import nn
from torch.nn import functional as F
from torch.utils.data import DataLoader, random_split, Dataset
from torchvision.datasets import MNIST
from torchvision import transforms
import pytorch_lightning as pl
from pytorch_lightning.metrics.functional import accuracy
tmpdir = os.getcwd()

from pl_bolts.datasets import RandomDataset, DummyDataset, RandomDictDataset
from pytorch_lightning.callbacks.early_stopping import EarlyStopping
from pytorch_lightning import loggers as pl_loggers

num_samples = 10000
train = RandomDataset(32, num_samples)
train = DataLoader(train, batch_size=32)
val = RandomDataset(32, num_samples)
val = DataLoader(val, batch_size=32)
test = RandomDataset(32, num_samples)
test = DataLoader(test, batch_size=32)

import torch
from pytorch_lightning import LightningModule
from torch.utils.data import Dataset

class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"loss": loss}

    def training_epoch_end(self, outputs) -> None:
        torch.stack([x["loss"] for x in outputs]).mean()

    def validation_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"x": loss}

    def predict_step(self, batch, batch_idx):
        output = self.layer(batch)
        return output

    def test_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        self.log('fake_test_acc', loss)
        return {"y": loss}

    def test_epoch_end(self, outputs) -> None:
        torch.stack([x["y"] for x in outputs]).mean()

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]

def test_x(tmpdir):

    # init model
    model = BoringModel()
    tb_logger = pl_loggers.TensorBoardLogger(save_dir="delay_logs/")
    NUM_GPUS = torch.cuda.device_count()
    EPOCHS = 15
    if NUM_GPUS > 1:
        trainer = pl.Trainer(precision=16, gpus=-1, strategy="ddp", max_epochs=EPOCHS, check_val_every_n_epoch=1,
                        logger = tb_logger, callbacks=[EarlyStopping(monitor="Val loss", patience=5)])
    else:
        trainer = pl.Trainer(gpus=1, max_epochs=EPOCHS, check_val_every_n_epoch=1,
                        logger = tb_logger, callbacks=[EarlyStopping(monitor="Val loss", patience=5)])

    # Train the model ⚡
    trainer.fit(model, train)

    trainer.test(test_dataloaders=test)

test_x(tmpdir)

```
-->

### Expected behavior

Nothing, happens, the process freezes and there is no progress. It shows the message ```NCCL backend is starting``` or it starts running, shows progress on the tqdm and freezes very soon (most of the time it doesn't even start)and if I check the GPU usage with ```nvidia-smi```, it shows GPU usage but the process continues for every without any output. If I change and make, ```strategy="dp"```, everything works smoothly.

### Environment

- Lightning Component: Trainer with strategy="ddp"
- PyTorch Lightning Version : 1.6.1
- PyTorch Version: 1.11
- Python version: 3.9.1
- OS: Debian 10
- Cluster Management: SLURM 
- CUDA/cuDNN version:  cuda_11.2.r11.2/compiler.29618528_0
- GPU models and configuration: Titan Xp
- How you installed PyTorch : pip
- Every node on the SLURM cluster has 8 GPUs, Titan Xp GPUs, 12 GB GPU memory, 28 core Dual Tetrakaideca-Core Xeon E5-2680 v4 @2.40 Ghz, 503 GB of RAM, 1TB Scratch 

### Additional context

To run on SLURM, I wrap my python file in the following Bash script
```
#!/bin/bash
#SBATCH --output=./%j.out
#SBATCH --error=./%j.err
#SBATCH --cpus-per-task=4
#SBATCH --gres=gpu:2
#SBATCH --mem=80G
set -o errexit  # Exit on errors

# Activate the correct venv.
eval "$(pyenv init -)"
eval "$(pyenv virtualenv-init -)"
pyenv activate venv

echo "Running on node: $(hostname)"
echo "In directory:    $(pwd)"
echo "Starting on:     $(date)"
echo "SLURM_JOB_ID:    ${SLURM_JOB_ID}"

NCCL_DEBUG=INFO; NCCL_DEBUG_SUBSYS=ALL python test.py <args>

```

Save this file above as ```run.sh``` and submit as ```sbatch run.sh```.

cc @awaelchli @tchaton @rohitgr7 @justusschock @kaushikb11 @akihironitta

**Comment by smartdanny** [comment]
I have the same issue. Mine hangs on Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/2

Ive tried with various numbers of gpu, nothing works.

**Comment by awaelchli** [comment]
Please have a look at our SLURM instructions [here](https://pytorch-lightning.readthedocs.io/en/stable/clouds/cluster_advanced.html). 
Pay close attention to the SBATCH variables. In particular, you are missing the setting `ntasks-per-node` and you are incorrectly launching the script with python. You should use srun instead.

**Comment by Siddhant-Ray** [comment]
Hi! Thanks for the reply. To launch the job, I save the above mentioned bash script as run.sh and then run using ```sbatch run.sh```, so launching the script is not the problem. I need to use ```sbatch``` to submit it to the scheduler.

My organisations SLURM docs didn't mention anything about the SBATCH variable ```ntasks-per-node``` but I will try it with that.

**Comment by awaelchli** [comment]
> there should be another srun command, so I will check that.

Yes exactly, I think you should replace

```
python test.py <args>
```

with
```
srun python test.py <args>
```

**Comment by Altrabeth** [comment]
Hi there, 

I'm having the same problem with SLURM and DDP, even when I follow all steps here and in the [wiki](https://lightning.ai/docs/pytorch/stable/clouds/cluster_advanced.html).

My sbatch script:

```bash
#!/bin/bash -l
#SBATCH --partition=single
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=2
#SBATCH --cpus-per-task=8
#SBATCH --gres=gpu:2
#SBATCH --mem-per-gpu=32Gb
#SBATCH --time=2:00:00
#SBATCH --output ./output/%A.out
#SBATCH --error ./error/%A.out

##### Number of total processes
echo "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX "
echo "Nodelist:= " $SLURM_JOB_NODELIST
echo "Number of nodes:= " $SLURM_JOB_NUM_NODES
echo "Ntasks per node:= "  $SLURM_NTASKS_PER_NODE
echo "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX "

source activate my_env
# ******************* These are read internally it seems ***********************************
# ******** Master port, address and world size MUST be passed as variables for DDP to work
export MASTER_PORT=$(expr 10000 + $(echo -n $SLURM_JOBID | tail -c 4))
export WORLD_SIZE=$(($SLURM_NNODES * $SLURM_NTASKS_PER_NODE))
echo "MASTER_PORT"=$MASTER_PORT
echo "WORLD_SIZE="$WORLD_SIZE

master_addr=$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)
export MASTER_ADDR=$master_addr
echo "MASTER_ADDR="$MASTER_ADDR
# ******************************************************************************************

srun python my_program.py
```

And my training setup in my_program.py:
```python
import logging
from argparse import Namespace
from logging import FileHandler
from pathlib import Path
from typing import Any

import pytorch_lightning as pl
import torch
import torchvision.models
from attr import define
from pytorch_lightning import Trainer
from pytorch_lightning.utilities.types import TRAIN_DATALOADERS, STEP_OUTPUT
from torch.optim import Adam
from torchvision.models import ResNet
from torchvision.transforms.v2 import (
    Compose,
    RandomCrop,
    RandomHorizontalFlip,
    RandomVerticalFlip,
    ToTensor,
)
import webdataset as wds
from webdataset import split_by_node
from pytorch_lightning.loggers import TensorBoardLogger

from propath.enums.trainer_stages import TrainerStage
from propath.enums.training_mode import TrainingMode
from propath.trainings.dataloader import DataLoader
from propath.trainings.learning_rate_schedulers.cosine_with_warmup import (
    CosineWithWarmupLR,
)
from propath.utility.argument_parsers import project_folder_argument_parser
from propath.utility.naming_pattern import tar_shard_range
from propath.utility.validators import type_check


class TestDataset(pl.LightningDataModule):
    def __init__(
        self,
        data_folder: Path,
        number_of_shards: int,
        epoch_size: int,
        batch_size_per_gpu: int,
        number_of_workers: int,
        prefetch_factor: int,
    ):
        super().__init__()

        type_check(data_folder, "data_folder", Path)
        type_check(number_of_shards, "number_of_shards", int)
        type_check(epoch_size, "epoch_size", int)
        type_check(batch_size_per_gpu, "batch_size_per_gpu", int)
        type_check(number_of_workers, "number_of_workers", int)
        type_check(prefetch_factor, "prefetch_factor", int)

        self.data_folder = data_folder
        self.number_of_shards = number_of_shards
        self.epoch_size = epoch_size
        self.batch_size_per_gpu = batch_size_per_gpu
        self.number_of_workers = number_of_workers
        self.prefetch_factor = prefetch_factor

    @staticmethod
    def make_data_transformation() -> Compose:
        return Compose(
            [RandomCrop(224), RandomHorizontalFlip(), RandomVerticalFlip(), ToTensor()]
        )

    def make_label_transformation(self) -> Compose:
        pass

    def setup(self, stage: TrainerStage) -> None:
        """This method expects a stage argument. It is used to separate setup logic for
        trainer.{fit,validate,test,predict}.

        Parameters
        ----------
        stage : TrainerStage
            Defines the mode of the Dataloader.
        """
        match stage:
            case TrainerStage.fit:
                dataset_path: Path = self.data_folder / tar_shard_range(
                    0, self.number_of_shards, TrainingMode.none
                )
                self.train_dataset: wds.WebDataset = (
                    wds.WebDataset(
                        dataset_path.as_posix(),
                        shardshuffle=True,
                        nodesplitter=split_by_node,
                    )
                    .decode("pil")
                    .to_tuple("png", "txt")
                    .map_tuple(
                        self.make_data_transformation(),
                        self.make_label_transformation(),
                    )
                    .with_length(self.epoch_size)
                    .with_epoch(self.epoch_size)
                )
            case TrainerStage.validate:
                raise NotImplementedError
            case TrainerStage.test:
                raise NotImplementedError
            case TrainerStage.predict:
                raise NotImplementedError
            case _:
                raise ValueError(f"Got unexpected value for training stage: {stage}.")

    def train_dataloader(self) -> TRAIN_DATALOADERS:
        return DataLoader(
            self.train_dataset,
            batch_size=self.batch_size_per_gpu,
            number_of_workers=self.number_of_workers,
            prefetch_factor=self.prefetch_factor,
        )


class LoggerNet(pl.LightningModule):
    def __init__(self, project_folder: Path, *args: Any, **kwargs: Any):
        super().__init__(*args, **kwargs)
        type_check(project_folder, "project_folder", Path)

        self.project_folder = project_folder

        self.net: ResNet = torchvision.models.resnet18()
        self.log_file_path: Path = project_folder / f"rank_{self.global_rank}.log"
        self.file_logging_handler = FileHandler(
            self.log_file_path,
        )
        formatter: logging.Formatter = logging.Formatter(
            "%(asctime)s %(levelname)s: %(message)s"
        )
        self.file_logging_handler.setFormatter(formatter)
        self.label_logger: logging.Logger = logging.getLogger()
        self.label_logger.addHandler(self.file_logging_handler)
        self.mse_loss = torch.nn.MSELoss()

    def forward(self, data):
        image, label = data
        return self.net(image)

    def configure_optimizers(self) -> Adam:
        optimizer: torch.optim.Adam = torch.optim.Adam(self.parameters(), lr=1e-3)
    def training_step(self, batch: torch.Tensor, batch_index: int) -> STEP_OUTPUT:
        data, labels = batch
        output: torch.Tensor = self.net(data)
        loss: torch.Tensor = self.mse_loss(output, output)
        self.label_logger.info(f"rank_{self.global_rank}_epoch_{self.current_epoch}_step_{self.global_step}_{labels}")
        self.log(
            "label_0",
            int(labels[0][7]) * 10 + int(labels[0][20]),
            on_step=True,
        )
        self.log(
            "label_1",
            int(labels[1][7]) * 10 + int(labels[1][20]),
            on_step=True,
        )
        return loss


@define
class ProjectPaths:
    project_folder: Path
    data_folder: Path
    lightning: Path


def get_project_paths(project_folder: str) -> ProjectPaths:
    project_folder: Path = Path(project_folder)
    return ProjectPaths(
        project_folder, project_folder / "data", project_folder / "lightning"
    )
def main():
    arguments: Namespace = project_folder_argument_parser()
    project_paths: ProjectPaths = get_project_paths(arguments.project_folder)
    logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

    data_model: TestDataset = TestDataset(project_paths.data_folder, 3, 40, 2, 4, 2)


    net: LoggerNet = LoggerNet(project_paths.project_folder)
    tensorboard_logger: TensorBoardLogger = TensorBoardLogger(save_dir=project_paths.project_folder)

    trainer: Trainer = Trainer(
        logger=tensorboard_logger,
        default_root_dir=project_paths.lightning,
        max_epochs=2,
        num_nodes=1,
        devices=2,
        strategy='ddp_spawn',
        accelerator="gpu",
    )

    trainer.fit(
        model=net,
        train_dataloaders=data_model,
    )


if __name__ == "__main__":
    main()
```

Please ignore the weird logging stuff in the code, I'm currently trying to reconstruct how the webdataset package handles multiple nodes.

Did I miss something?

**Comment by awaelchli** [comment]
In the python code, `strategy='ddp_spawn'` sticks out to me. Set this to "ddp" or remove the line (it will auto-detect). 
In the slurm script, remove these lines:

```
# ******************* These are read internally it seems ***********************************
# ******** Master port, address and world size MUST be passed as variables for DDP to work
export MASTER_PORT=$(expr 10000 + $(echo -n $SLURM_JOBID | tail -c 4))
export WORLD_SIZE=$(($SLURM_NNODES * $SLURM_NTASKS_PER_NODE))
echo "MASTER_PORT"=$MASTER_PORT
echo "WORLD_SIZE="$WORLD_SIZE

master_addr=$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)
export MASTER_ADDR=$master_addr
echo "MASTER_ADDR="$MASTER_ADDR
```
The tutorial doesn't say anywhere that these should be set, this tells me that you didn't follow the steps 😉

**Comment by Altrabeth** [comment]
I'm sorry. I uploaded the latest version of my code. I tried 'ddp', 'dp' (which failed to my surprise) and 'ddp_spawn', which also failed.

**Comment by Altrabeth** [comment]
I updated the code as descibed:

```bash
#!/bin/bash -l
#SBATCH --partition=single
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=2
#SBATCH --gres=gpu:2
#SBATCH --mem-per-gpu=32Gb
#SBATCH --time=2:00:00
#SBATCH --output ./output/%A.out
#SBATCH --error ./error/%A.out

echo "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX "
echo "Nodelist:= " $SLURM_JOB_NODELIST
echo "Number of nodes:= " $SLURM_JOB_NUM_NODES
echo "Ntasks per node:= "  $SLURM_NTASKS_PER_NODE
echo "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX "

source activate my_env

srun python my_program.py --project_folder $1
```

```python
 # ... see above ...

def main():
    print("start")
    arguments: Namespace = project_folder_argument_parser()
    project_paths: ProjectPaths = get_project_paths(arguments.project_folder)
    logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
    print("parser")
    data_model: TestDataset = TestDataset(project_paths.data_folder, 3, 40, 2, 4, 2)
    print("net")

    net: LoggerNet = LoggerNet(project_paths.project_folder)
    tensorboard_logger: TensorBoardLogger = TensorBoardLogger(save_dir=project_paths.project_folder)
    print("trainer")
    trainer: Trainer = Trainer(
        logger=tensorboard_logger,
        default_root_dir=project_paths.lightning,
        max_epochs=2,
        num_nodes=1,
        devices=2,
        strategy='ddp',
        accelerator="gpu",
    )
    print("fit")
    trainer.fit(
        model=net,
        train_dataloaders=data_model,
    )
    

   # ... see above ...
```

Sadly, the behavior stays the same. It still performs the first iteration and then stops.

And I have an additional question. If I don't set `#SBATCH --cpus-per-task=8` how do I tell the process how many CPU cores it should use (for data loader processes, which would be in my example 7)?
